### PR TITLE
issue #2184 - support getting all the common token value ids in one go

### PR DESCRIPTION
**Pull request by lmsurpre**
Added:
* getCommonTokenValueIds to JDBCIdentityCache
* resolveCommonTokenValueIds to ICommonTokenValuesCache
* readCommonTokenValueIds to ResourceReferenceDAO

Updated SearchQueryRenderer to use this mechanism in two cases:
* getReferenceFilter for reference params without the `:identifier` modifier (including compartment params)
* getTokenFilter for token params with the IN, ABOVE, BELOW, and NOT-IN modifiers

Also includes cleanup in getReferenceFilter and populateCodesSubSegment (used by getTokenFilter):
* removal of code in getReferenceFilter for handling modifier that we don't support for reference params
* removal of code that uses the token ref view for token filters...I believe that the IN, ABOVE, BELOW, and NOT-IN cases can be handled with COMMON_TOKEN_VALUES now

Needs testing on Db2.

### CODE REVIEW

File: fhir-database-utils/src/main/java/com/ibm/fhir/database/utils/query/expression/StringStatementRenderer.java [hunk 1]
```diff
@@ -92,7 +92,7 @@ public String select(boolean distinct, SelectList selectList, FromClause fromCla
 
         if (groupByClause != null) {
             if (this.pretty) {
-                result.append(NEWLINE).append("    ");
```
Comment (lmsurpre) [inline fhir-database-utils/src/main/java/com/ibm/fhir/database/utils/query/expression/StringStatementRenderer.java:95]: this is a minor formatting fix that is unrelated to the rest of the PR

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/derby/DerbyResourceReferenceDAO.java [hunk 2]
```diff
@@ -157,6 +191,29 @@ public void doCanonicalValuesUpsert(String paramList, Collection<String> urls) {
...
+        try (Statement s = getConnection().createStatement()) {
+            s.executeUpdate(upsert.toString());
+        } catch (SQLException x) {
+            logger.log(Level.SEVERE, upsert.toString(), x);
+            throw getTranslator().translate(x);
+        }
+    }
+
+    private void insertToCommonTokenValuesTmp(Collection<CommonTokenValue> tokenValues) {
         final String insert = "INSERT INTO SESSION.common_token_values_tmp(token_value, code_system_id) VALUES (?, ?)";
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/derby/DerbyResourceReferenceDAO.java:217]: I wasn't totally sure if it was safe to use the same temporary table name here for both doCommonTokenValuesUpsert and the new readCommonTokenValueIds.

In the latest CI run I'm seeing errors like the following:
```
[ERROR   ] INSERT INTO common_token_values (token_value, code_system_id)      SELECT src.token_value, src.code_system_id        FROM SESSION.common_token_values_tmp src  LEFT OUTER JOIN common_token_values ctv               ON ctv.token_value = src.token_value              AND ctv.code_system_id = src.code_system_id       WHERE ctv.token_value IS NULL    ORDER BY src.token_value, src.code_system_id
A lock could not be obtained within the time requested
```

So I'm thinking to have this shared helper accept the name for the temporary table as an argument and then ensure that the two different method pass a unique value.

Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/derby/DerbyResourceReferenceDAO.java:224]: scratch that, I went with the simpler `WHERE (c.token_value = ? AND c.code_system_id = ?) OR (c.token_value = ? AND c.code_system_id = ?) OR ...` approach instead because it just felt safer to me and I don't want to spend too much time optimizing for derby

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/derby/DerbyResourceReferenceDAO.java [hunk 3]
```diff
@@ -49,6 +53,43 @@ public DerbyResourceReferenceDAO(IDatabaseTranslator t, Connection c, String sch
...
 
+    @Override
+    public Set<CommonTokenValueResult> readCommonTokenValueIds(Collection<CommonTokenValue> tokenValues) {
+        if (tokenValues.isEmpty()) {
+            return Collections.emptySet();
+        }
+
+        Set<CommonTokenValueResult> result = new HashSet<>();
+
+        String SQL = ""
```
Comment (punktilious) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/derby/DerbyResourceReferenceDAO.java:64]: Should use StringBuilder here

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/ResourceReferenceDAO.java [hunk 4]
```diff
@@ -152,6 +155,49 @@ public CommonTokenValueResult readCommonTokenValueId(String codeSystem, String t
...
 
+    @Override
+    public Set<CommonTokenValueResult> readCommonTokenValueIds(Collection<CommonTokenValue> tokenValues) {
+        if (tokenValues.isEmpty()) {
+            return Collections.emptySet();
+        }
+
+        Set<CommonTokenValueResult> result = new HashSet<>();
+
+        String SQL = ""
```
Comment (punktilious) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/ResourceReferenceDAO.java:166]: StringBuilder

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java [hunk 5]
```diff
@@ -522,8 +537,45 @@ protected WhereFragment getTokenFilter(QueryParameter queryParm, String paramAli
...
+                //
+                //
+                // Scenario 2: none of the resources have any of the values from the valueset
+                //
+                //    WHERE NOT EXISTS(
+                //    SELECT 1
+                //    FROM Basic_RESOURCE_TOKEN_REFS AS P2
+                //    WHERE P2.LOGICAL_RESOURCE_ID = LR1.LOGICAL_RESOURCE_ID
+                //      AND P2.PARAMETER_NAME_ID = 20046
+                //      AND ((P2.COMMON_TOKEN_VALUE_ID IN (-1)))))) AS LR
```
Comment (punktilious) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java:571]: Just a thought...if we only have -1 in the set here, it might be worth replacing that with NULL. This _may_ let the database optimize the query more, because it knows that NULL will never match anything and therefore it could short-circuit the join (if it's clever enough). With -1, it still has to go fishing in the index to find that value. May also impact cardinality estimation (hopefully in a positive way).

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/CommonTokenValue.java [hunk 6]
```diff
@@ -11,18 +11,18 @@
...
-    
+
     private final int codeSystemId;
-    
+
     // tokenValue can be null
     private final String tokenValue;
-    
+
     public CommonTokenValue(int codeSystemId, String tokenValue) {
```
Comment (punktilious) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/CommonTokenValue.java:20]: Missing javadoc here (my bad)

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/JDBCIdentityCacheImpl.java [hunk 7]
```diff
@@ -160,6 +164,38 @@ public Long getCommonTokenValueId(String codeSystem, String tokenValue) {
...
+        if (misses.isEmpty()) {
+            return result;
+        }
+
+        if (logger.isLoggable(Level.FINE)) {
+            logger.fine("Cache miss. Fetching common_token_value_ids from database: " + misses);
+        }
+
+        Set<CommonTokenValueResult> readCommonTokenValueIds = resourceReferenceDAO.readCommonTokenValueIds(misses);
+        result.addAll(readCommonTokenValueIds.stream()
```
[the remark this answers is not in the document]
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/JDBCIdentityCacheImpl.java:181]: just that it was a tad slower in code with very tight loops.  I don't think we need to avoid ALL uses of streams because of that.

[the remark this answers is not in the document]
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/JDBCIdentityCacheImpl.java:181]: this particular method should happen once per search parameter and typically we don't have a lot of those.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/CommonTokenValue.java (hunk 6 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/CommonTokenValue.java:25]: added javadoc

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java (hunk 5 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java]: Its actually a bind variable (I had simplified it for my comment).  I went ahead and changed the behavior so that we bind a null instead of a -1 (at least for the case of token params).
For reference params, we use BaseWhereAdapter.eq which takes primitive values, so in that case I'm still using the `-1`.

We should probably resolve on a consistent approach to these.  Is it better to use literals?  Is it better to use null?

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/JDBCIdentityCacheImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/JDBCIdentityCacheImpl.java:181]: Yeah, just thought I'm mention it in case you thought it could have an impact. It's a shame about the performance, because the syntactic convenience is nice. Agree we probably don't have to worry unless it shows up in a profile down the road.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java (hunk 5 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java]: Comparisons with common_token_value_id shouldn't use bind variables - we always want to use literals to make sure we get the best plan. With PostgreSQL it's more interesting, because it definitely peeks more at the bind values when optimizing, but other databases may not. Anyway, our standard approach is to always use literals where it's a database id (where there is little chance the query will be reused).

As for OR vs IN list, the optimizer will usually automatically convert...but I tend to prefer an IN list because it's neater.

If you *only* have -1 - e.g. WHERE foo_id IN (-1) then we should rewrite that to be WHERE foo_id IN (NULL). That gives the optimizer some useful information.

**Review by punktilious — APPROVED** [review]
LGTM

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java (hunk 5 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java]: Our current BaseWhereAdapter doesn't seem to support the use of NULL literals (in either eq or in expressions), so unless we change that we need to pick whether to use literals or whether to use NULLs.  For now, I opted to use literals across these two methods now and we can use #2688 if we want to add support for replacing the -1 values with NULLs.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/CommonTokenValue.java [hunk 8]
```diff
@@ -11,18 +11,23 @@
  * DTO representing a record in COMMON_TOKEN_VALUES
```
Comment (tbieste) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/CommonTokenValue.java:11]: Update copyright.

**Review by tbieste — APPROVED** [review]
Left one minor comment.  Otherwise, LGTM.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/CommonTokenValue.java (hunk 8 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dto/CommonTokenValue.java:11]: done

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java (hunk 5 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java:570]: For completeness, I just wanted to update here and say we should probably use `WHERE foo_id = NULL` for the case where there is a single value. That probably gives us the best chance that the optimizer will see it and short-circuit the join (because nothing ever is equal to NULL). Of course, if we know this we could probably rewrite the query ourselves, but that is complicated, so we are delegating responsibility to the DB layer, in the hope it will do the right thing.
